### PR TITLE
FLUX.2 Improvements and fixes

### PR DESCRIPTION
**Pull request by kohya-ss**
Implement (items will be added):
- [x] fp8 optimization with LoRA loading
- [x] Negative prompt for base on generation
- [x] Support unified attention mechanism (xformers, flash attention, sage attention)
- [x] Support higher batch size training if possible for training with control images
- [x] Split uncond and cond in CFG to reduce memory usage
- [x] Move control image resizing from caching script to dataset script
- [x] Different architecture suffixes for dev/klein 4B/klein 9B due to different Text Encoders
- [x] Refactor to use Aspect Ratio Bucketing resizing to limit total pixel size of the image for generation
- [x] Update block calculation for block swap based on the ratio of double/single blocks (because double*2≠single)
- [x] Generalize qwen_imag_edit/flux_kontext_no_resize_control and qwen_image_edit_control_resolution to use them with FLUX.2
- [x] To cap control image size, allow no_resize_control and control_resolution to be used simultaneously
- [x] Refactor sample image generation in training same as generation
- [x] Model optimization

Verification tasks:
- [x] Sequence length of text embeddings (always use full 512 or apply attention mask?) -> use 512 length
- [x] VAE dtype (float32 only or bfloat16 works) -> VAE seems to work with bfloat16. Make float32 default and bfloat16 optional.

Tests:
- [x] klein 4B generation
- [x] klein 4B caching
- [x] klein 4B training
- [x] klein 9B generation
- [x] klein 9B caching
- [x] ~~klein 9B training~~
- [x] Generation with control images
- [x] Caching with control images
- [x] Training with control images

### CODE REVIEW

**Comment by kohya-ss** [comment]
Musubi Tuner performs aspect ratio bucketing on the dataset, including the reference image size, so training with batch sizes greater than 1 is also possible.

Note that since the outputs of the Text Encoder are all the same length, if we concatenate text, generated image, and reference image in the order and create an attention mask based on the length of the reference image, we can include different reference image sizes and numbers in the same batch. However, it is probably not worth implementing this.

**Comment by recris** [comment]
Is it possible to set a different text embedding length as an input? If the captions in the training data don't need that many tokens, then reducing the embedding length can improve performance.

For example, training Flux.1 lora (with `sd-scripts` repo) but reducing length from 512 to 128 yields a 20% reduction in training step time, using an AMD GPU. I haven't observed a worse quality output from doing this.

**Comment by kohya-ss** [comment]
> Is it possible to set a different text embedding length as an input? If the captions in the training data don't need that many tokens, then reducing the embedding length can improve performance.

I think that's possible, and I'll address that in another PR once this one is complete.

File: docs/flux_2.md [hunk 1]
```diff
@@ -17,18 +17,33 @@ Latent pre-caching, training, and inference options can be found in the `--help`
...
+
+### FLUX.2 [klein] 9B / base 9B
+
+- **DiT 9B**: Download from the [black-forest-labs/FLUX.2-klein-9B](https://huggingface.co/black-forest-labs/FLUX.2-klein-9B) repository. Use `flux2-klein-9b.safetensors`.
+- **DiT base 9B**: Download from the [black-forest-labs/FLUX.2-klein-base-9B](https://huggingface.co/black-forest-labs/FLUX.2-klein-base-9B) repository. Use `flux2-klein-base-9b.safetensors`.
+- **AE**: Download from the [black-forest-labs/FLUX.2](https://huggingface.co/black-forest-labs/FLUX.2-dev) repository. Use `ae.safetensors`. `vae/diffusion_pytorch_model.safetensors` in the subfolder is in Diffusers format and cannot be used.
+- **Qwen3 8B Text Encoder**: Download all the split files from the [black-forest-labs/FLUX.2-klein-9B](https://huggingface.co/black-forest-labs/FLUX.2-klein-9B) repository and specify the first file (e.g., `00001-of-00004.safetensors`) in the arguments.
 
 ## Specifying Model Version / モデルバージョンの指定
```
Comment (tokinasin) [inline docs/flux_2.md:72]: It seems like some model names and their corresponding parameters are misaligned in the table.

Comment (kohya-ss) [reply docs/flux_2.md:72]: The klein repositories only have the Diffusers version of AE, so we need to use AE in dev repository.

Comment (tokinasin) [reply docs/flux_2.md:72]: Hi! I meant to point out the misalignment in the table here:

https://github.com/kohya-ss/musubi-tuner/blob/5768207e27fbd47dd5b67bec15c39ccd5ded5478/docs/flux_2.md?plain=1#L53-L57

Specifically, the parameters seem mixed up (e.g. the standard flux.2-klein (non-base) models are listed with scale 4.0 and 50 steps, but i think they should be 1.0 and 4 steps.)

sorry fot the confusion m(__)m

File: src/musubi_tuner/flux_2_cache_latents.py [hunk 2]
```diff
@@ -26,36 +21,32 @@ def preprocess_contents_flux_2(batch: List[ItemInfo]) -> tuple[torch.Tensor, Lis
...
-            else:
-                limit_pixels = None
-            img_ctx = [(Image.fromarray(cc) if isinstance(cc, np.ndarray) else cc).convert("RGB") for cc in item.control_content]
-
-            img_ctx_prep = flux2_utils.default_prep(
-                img=img_ctx,
-                limit_pixels=limit_pixels,
-            )
-            controls.append(img_ctx_prep)
+    contents = torch.stack(contents, dim=0)  # B, H, W, C
```
Comment (recris) [inline src/musubi_tuner/flux_2_cache_latents.py]: This line is duplicated with the next one, and causes an error.

File: src/musubi_tuner/flux_2/flux2_utils.py [hunk 3]
```diff
@@ -39,41 +54,62 @@
 attribution and actions without speculation."""
 
 
+@dataclass(frozen=True)
+class Flux2ModelInfo:
+    params: Flux2Params
+    defaults: dict[str, float | int]
+    fixed_params: set[str]
+    guidance_distilled: bool
+    qwen_variant: Optional[str] = None  # None for Mistral
```
Comment (recris) [inline src/musubi_tuner/flux_2/flux2_utils.py]: The variable with default value needs to come last, this causes a `TypeError`

**Comment by recris** [comment]
Just ran a test using Klein Base 9B - aside from a couple of minor changes (see comments) it seems to have produced a working LoRA.

**Comment by otherroch** [comment]
The other minor problem related to training is:

```
> @@ -56,10 +56,11 @@ class Flux2NetworkTrainer(NetworkTrainer):
>          logger.info(f"cache Text Encoder outputs for sample prompt: {sample_prompts}")
>          prompts = load_prompts(sample_prompts)
> 
> +        model_version_info = flux2_utils.FLUX2_MODEL_INFO[args.model_version]
>          # Load Text Encoder (Mistral 3 or Qwen-3)
>          te_dtype = torch.float8e4m3fn if args.fp8_text_encoder else torch.bfloat16
>          text_embedder = flux2_utils.load_text_embedder(
> -            args.model_version, args.text_encoder, dtype=te_dtype, device=device, disable_mmap=True
> +            model_version_info, args.text_encoder, dtype=te_dtype, device=device, disable_mmap=True
>          )
> 
>          # Encode with Text Encoder (Mistral 3 or Qwen-3)
```

File: docs/flux_2.md (hunk 1 above)
Comment (kohya-ss) [reply docs/flux_2.md:72]: Thank you! I've fixed.

File: src/musubi_tuner/flux_2_cache_latents.py (hunk 2 above)
Comment (kohya-ss) [reply src/musubi_tuner/flux_2_cache_latents.py]: Thank you, I've updated.

File: src/musubi_tuner/flux_2/flux2_utils.py (hunk 3 above)
Comment (kohya-ss) [reply src/musubi_tuner/flux_2/flux2_utils.py]: Thanks, this is also fixed.

**Comment by kohya-ss** [comment]
> The other minor problem related to training is:

Thank you, this is fixed.

File: docs/flux_2.md (hunk 1 above)
Comment (tokinasin) [reply docs/flux_2.md:72]: flux.2-dev may also need to be 4.0 and 50.

Comment (kohya-ss) [reply docs/flux_2.md:72]: Oh, thank you again. dev is distilled but needs 50 steps and embedded guidance scale=4.0.

Comment (kohya-ss) [reply docs/flux_2.md:72]: I've updated :)

**Comment by kohya-ss** [comment]
I have confirmed that using the `--xformers` or `--flash_attn` options instead of the `--sdpa` option also works. Both options reduce the training time by about 25%.

File: src/musubi_tuner/flux_2_generate_image.py [hunk 4]
```diff
@@ -602,47 +662,39 @@ def generate(
...
+    logger.info(f"Prompt: {context['prompt']}, Negative Prompt: {context_null['prompt'] if context_null is not None else 'N/A'}")
     ctx_vec = context["ctx_vec"].to(device, dtype=torch.bfloat16)
-    ctx, ctx_ids = flux2_utils.batched_prc_txt(ctx_vec)
+    ctx, ctx_ids = flux2_utils.prc_txt(ctx_vec)
+    if context_null is None:
+        negative_ctx_vec = None
+        ctx_null, ctx_null_ids = None, None
+    else:
+        negative_ctx_vec = context_null["ctx_vec"].to(device, dtype=torch.bfloat16)
+        ctx_null, ctx_null_ids = flux2_utils.prc_txt(negative_ctx_vec)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/flux_2_generate_image.py:657]: Keep the arg for readability.

File: src/musubi_tuner/flux_2_cache_text_encoder_outputs.py [hunk 5]
```diff
@@ -20,12 +16,7 @@
...
 
-def encode_and_save_batch(
-    text_embedder: torch.nn.Module,
-    guidance_distilled: bool,
-    batch: list[ItemInfo],
-    device: torch.device,
-):
+def encode_and_save_batch(text_embedder: torch.nn.Module, batch: list[ItemInfo], device: torch.device, arch_full: str):
     prompts = [item.caption for item in batch]
     with torch.autocast(device_type=device.type, dtype=text_embedder.dtype), torch.no_grad():
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/flux_2_cache_text_encoder_outputs.py]: I'll fix this.

File: src/musubi_tuner/flux_2/flux2_utils.py [hunk 6]
```diff
@@ -367,14 +350,10 @@ def denoise(
...
-        pred = model(
-            x=img_input,
-            x_ids=img_input_ids,
-            timesteps=t_vec,
-            ctx=txt,
-            ctx_ids=txt_ids,
-            guidance=guidance_vec,
-        )
+
+        with torch.no_grad(), torch.autocast(device_type=img.device.type, dtype=model.dtype):
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/flux_2/flux2_utils.py]: img.dtype is bfloat16, so it's fine to use img.dtype here. I'll update this.

File: src/musubi_tuner/flux_2/flux2_utils.py [hunk 7]
```diff
@@ -417,25 +390,20 @@ def denoise_cfg(
...
-            timesteps=t_vec,
-            ctx=txt,
-            ctx_ids=txt_ids,
-            guidance=None,
-        )
+        with torch.no_grad(), torch.autocast(device_type=img.device.type, dtype=model.dtype):
+            pred_cond = model(x=img_input, x_ids=img_input_ids, timesteps=t_vec, ctx=txt, ctx_ids=txt_ids, guidance=None)
+            pred_uncond = model(
+                x=img_input, x_ids=img_input_ids, timesteps=t_vec, ctx=uncond_txt, ctx_ids=uncond_txt_ids, guidance=None
+            )
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/flux_2/flux2_utils.py]: I'll update to use img.dtype instead of model.dtype.

File: src/musubi_tuner/flux_2/flux2_utils.py [hunk 8]
```diff
@@ -393,22 +372,16 @@ def denoise_cfg(
...
-    img_ids = torch.cat([img_ids, img_ids], dim=0)
-
-    if img_cond_seq is not None:
-        assert img_cond_seq_ids is not None
-        img_cond_seq = torch.cat([img_cond_seq, img_cond_seq], dim=0)
-        img_cond_seq_ids = torch.cat([img_cond_seq_ids, img_cond_seq_ids], dim=0)
-
-    for t_curr, t_prev in zip(timesteps[:-1], timesteps[1:]):
+    for t_curr, t_prev in zip(tqdm(timesteps[:-1]), timesteps[1:]):
         t_vec = torch.full((img.shape[0],), t_curr, dtype=img.dtype, device=img.device)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/flux_2/flux2_utils.py:385]: It's fine to show progress bar in sample generation. I'll consider this in future.

File: src/musubi_tuner/flux_2/flux2_utils.py [hunk 9]
```diff
@@ -270,59 +264,48 @@ def cap_pixels(img: Image.Image | list[Image.Image], k):
...
+            control_image.size, control_image.size, architecture=ARCHITECTURE_FLUX_2_DEV
+        )
     else:
-        img_cap = img_min
-    img_crop = center_crop_to_multiple_of_x(img_cap, ensure_multiple)  # type: ignore
-    img_tensor = default_images_prep(img_crop)
-    return img_tensor
+        resize_size = limit_size
+
+    control_image_tensor, control_image_np, _ = image_utils.preprocess_image(control_image, *resize_size, handle_alpha=False)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/flux_2/flux2_utils.py:294]: I'll update this.

**Comment by kohya-ss** [comment]
Although it has only been tested with klein-base-4B, it is likely to work with 9B as well, so we will merge this.

Note that training with multiple control images has not been tested.

**Comment by kohya-ss** [comment]
It seems that the trained LoRA can be used with ComfyUI.

**Comment by BenDes21** [comment]
Hi, is Klein 9b can be full finetuned ? Thanks
